### PR TITLE
Reimplement drag n drop

### PR DESCRIPTION
**Pull request by Nachtalb**
- Reimplement the drag n drop feature
- Add journal entries (with name)
- Works on plone 4 and 5
- Improved styles and usability compared to before
- Cancelling and upload by hitting ESC now actually works
- Implementation based on a0f9c481a815979f378bf14bfa2089a868eaa70c

https://4teamwork.atlassian.net/browse/OGB-702

Upload and add a journal entry
---

Cancel upload with ESC
---

### CODE REVIEW

File: ftw/file/browser/file_view.pt [hunk 1]
```diff
@@ -14,6 +14,30 @@
...
+        </div>
+
+        <p tal:condition="view/metadata/can_edit" id="dnd-file-replacement-hint" i18n:translate="">Hint: Drag and drop a file from your computer into the browser in order to replace the file.</p>
+
+        <!-- TAL expressions are not evaluated inside script tags. -->
+        <script tal:content="string:var context_url = '${context/absolute_url}'"></script>
+        <script>
+          (function (root, factory) {
+            if (typeof define === 'function' && define.amd) {
+              // AMD. Register as an anonymous module.
```
Comment (djowett-ftw) [inline ftw/file/browser/file_view.pt:31]: This ain't a module matey :-)

**Review by djowett-ftw — COMMENTED** [review]
Since dropzone.js is included in Plone 5,  I really think we are missing an opportunity here to get rid of legacy JavaScript code as we move forward with this product - and reduce our technical debt.  Do you want to being supporting old style JS unnecessarily?
  If we need to support Plone 4, then granted, that's something to weigh up, but I think it's worth mentioning anyway.

I've not looked at it much, but it might be mostly a simple template thing using pat-upload.... 

```
    <div class="pat-upload" data-pat-upload='{"url": "/ajax-upload"}'>
       <p>Upload here</p>
    </div>
```
I also noticed that file_download_version fails in the Plone 4 history - but that may not be the fault of this patch

File: ftw/file/profiles/default/repositorytool.xml [hunk 2]
```diff
@@ -2,7 +2,6 @@
 <repositorytool>
   <policymap>
     <type name="ftw.file.File">
-      <policy name="at_edit_autoversion"/>
```
[the remark this answers is not in the document]
Comment (Nachtalb) [reply ftw/file/profiles/default/repositorytool.xml]: Removed that in https://github.com/4teamwork/ftw.file/commit/d61f6eb740fdf2de5d9066ad0686d608fcc5712d

File: ftw/file/browser/file_view.pt (hunk 1 above)
Comment (Nachtalb) [reply ftw/file/browser/file_view.pt:31]: tbh that was copy-paste tbh

File: ftw/file/browser/upload.py [hunk 3]
```diff
@@ -4,14 +4,25 @@
...
+    DISABLE_CSRF = False
+
 
 class FileUpload(BrowserView):
     """View for handling drag-and-drop file replace"""
 
     def __call__(self):
+        if DISABLE_CSRF:
+            # Necessary for plone 5
+            alsoProvides(self.request, IDisableCSRFProtection)
```
[the remark this answers is not in the document]
Comment (maethu) [reply ftw/file/browser/upload.py:24]: I agree i should be easy to add CSRF Protection here by adding the hidden input field on the detail view + use it in the POST/PATCH request.

**Comment by maethu** [comment]
IMHO it looks good beside the disables CSRF Protection.

**Comment by Nachtalb** [comment]
After discussing: we won't use the dropzone pattern from Plone 5. Like that we would either need to have to separate implementations (once Plone 5, once 4) or remove the feature in Plone 4. https://4teamwork.slack.com/archives/C0D69FKRD/p1598942668039500

**Comment by Nachtalb** [comment]
@maethu @djowett-ftw I added CSRF protection as well as hardened the tests

**Review by djowett-ftw — CHANGES_REQUESTED** [review]
I get this on Plone5 (haven't tested Plone 4)

 ```Uncaught TypeError: e(...).overlay is not a function
    c ftw-file-compiled.js:154
    jQuery 5
    <anonymous> ftw-file-compiled.js:196
    execCb require.js:1696
    check require.js:883
    enable require.js:1176
    init require.js:788
    localRequire require.js:1460
    setTimeout handler*req.nextTick< require.js:1815
    localRequire require.js:1449
    requirejs require.js:1797
    <anonymous> ftw-file-compiled.js:93
    <anonymous> ftw-file-compiled.js:95
ftw-file-compiled.js:154
    c ftw-file-compiled.js:154
    jQuery 5
    <anonymous> ftw-file-compiled.js:196
    execCb require.js:1696
    check require.js:883
    enable require.js:1176
    init require.js:788
    localRequire require.js:1460
    (Async: setTimeout handler)
    nextTick require.js:1815
    localRequire require.js:1449
    requirejs require.js:1797
    <anonymous> ftw-file-compiled.js:93
    <anonymous> ftw-file-compiled.js:95
```

File: ftw/file/resources/uploader.js [hunk 4]
```diff
@@ -114,8 +114,9 @@
       }
     },
     init = function() {
-      $.event.props.push("dataTransfer");
       dropzone = document.getElementById('dropzone');
+      if (dropzone === null) return;
```
[the remark this answers is not in the document]
Comment (maethu) [reply ftw/file/resources/uploader.js:118]: Same with the SCSS.

File: ftw/file/resources/uploader.js [hunk 5]
```diff
@@ -28,6 +28,7 @@
       dropzone.className = 'uploading';
       if (tests.formdata) {
         formData.append('file', file);
+        formData.append('_authenticator', $('#dropzone [name="_authenticator"').val());
```
[the remark this answers is not in the document]
Comment (Nachtalb) [reply ftw/file/resources/uploader.js:31]: Damn my rebase failed. I had already fixed that but have lost it somehow 😂 Thanks!

**Comment by Nachtalb** [comment]
@maethu as discussed I now fixed the download and tested it on plone 4 and 5

**Review by maethu — APPROVED** [review]
💚  AWESOME

**Comment by maethu** [comment]
@Nachtalb BUT ADD A CHANGELOG ENTRY 😆

**Comment by djowett-ftw** [comment]
> I get this on Plone5 (haven't tested Plone 4)
> 
> ```
>    c ftw-file-compiled.js:154
>    jQuery 5
>    <anonymous> ftw-file-compiled.js:196
>    execCb require.js:1696
>    check require.js:883
>    enable require.js:1176
>    init require.js:788
>    localRequire require.js:1460
>    setTimeout handler*req.nextTick< require.js:1815
>    localRequire require.js:1449
>    requirejs require.js:1797
>    <anonymous> ftw-file-compiled.js:93
>    <anonymous> ftw-file-compiled.js:95
> ftw-file-compiled.js:154
>    c ftw-file-compiled.js:154
>    jQuery 5
>    <anonymous> ftw-file-compiled.js:196
>    execCb require.js:1696
>    check require.js:883
>    enable require.js:1176
>    init require.js:788
>    localRequire require.js:1460
>    (Async: setTimeout handler)
>    nextTick require.js:1815
>    localRequire require.js:1449
>    requirejs require.js:1797
>    <anonymous> ftw-file-compiled.js:93
>    <anonymous> ftw-file-compiled.js:95
> ```

@Nachtalb Just checking - you did see this comment, right? I think it's fixed now on firefox, but would be good to hear you think so too :-)

**Comment by djowett-ftw** [comment]
Sorry - that wasn't a good stacktrace I think.

This is the one I saw - which I just had again on https://devizug.zg.ch/web/behoerden/test/various-files/dsc_0210.jpg/file_view
```
Uncaught TypeError: e(...).overlay is not a function
    p https://devizug.zg.ch/web/++plone++production/++unique++2020-09-03T16:17:44.662058/default.js:11877
    j https://devizug.zg.ch/web/++plone++production/++unique++2020-09-03T16:17:44.662058/default.js:4
    add https://devizug.zg.ch/web/++plone++production/++unique++2020-09-03T16:17:44.662058/default.js:4
    ready https://devizug.zg.ch/web/++plone++production/++unique++2020-09-03T16:17:44.662058/default.js:4
    init https://devizug.zg.ch/web/++plone++production/++unique++2020-09-03T16:17:44.662058/default.js:4
    m https://devizug.zg.ch/web/++plone++production/++unique++2020-09-03T16:17:44.662058/default.js:4
    <anonymous> https://devizug.zg.ch/web/++plone++production/++unique++2020-09-03T16:17:44.662058/default.js:11877
    execCb https://devizug.zg.ch/web/++plone++production/++unique++2020-09-03T16:17:44.662058/default.js:1707
    check https://devizug.zg.ch/web/++plone++production/++unique++2020-09-03T16:17:44.662058/default.js:894
    enable https://devizug.zg.ch/web/++plone++production/++unique++2020-09-03T16:17:44.662058/default.js:1187
    init https://devizug.zg.ch/web/++plone++production/++unique++2020-09-03T16:17:44.662058/default.js:799
    localRequire https://devizug.zg.ch/web/++plone++production/++unique++2020-09-03T16:17:44.662058/default.js:1471
default.js:11877:4411
```

Screenshot 2020-09-09 at 16 29 17

**Comment by Nachtalb** [comment]
@djowett-ftw fixed with https://github.com/4teamwork/ftw.file/pull/117

**Pull request merged**